src/main/java/cn/RecipeAPI/Security/CustomUserDetailsService.java

```diff
@@ -0,0 +1,79 @@
+package cn.RecipeAPI.Security;
+
+import cn.RecipeAPI.Models.CustomUserDetails;
+import cn.RecipeAPI.Models.Role;
+import cn.RecipeAPI.Repositories.UserRepo;
+import org.hibernate.Hibernate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.persistence.EntityNotFoundException;
+import java.util.Collections;
+
+@Component
+public class CustomUserDetailsService implements UserDetailsService {
+
+    @Autowired
+    UserRepo userRepo;
+
+    @Autowired
+    PasswordEncoder encoder;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        CustomUserDetails optionalUser = userRepo.findByUsername(username);
+
+        if (optionalUser == null) {
+            throw new UsernameNotFoundException(username + " is not a valid username! Check for typos and try again.");
+        }
+
+        return optionalUser;
+    }
+
+    @Transactional(readOnly = true)
+    public CustomUserDetails getUserByUserId(Long userId) throws EntityNotFoundException {
+        CustomUserDetails user = userRepo.getReferenceById(userId);
+
+        //call unproxy() to ensure all related entities are loaded—no lazy load exceptions.
+        return (CustomUserDetails) Hibernate.unproxy(user);
```

Comment:
Cool!

Reply:
Ha! Yeah, that's Ryan's code. I don't actually understand what's happening on line 43.